File: API/PageAndRDL/ReportWizard/C#/ReportWizard/UI/DragDropListBox.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
namespace GrapeCity.ActiveReports.Samples.ReportWizard.UI 
{
	public class DragDropListBox : ListBox
	{
		public class DataObjectEventArgs : EventArgs
		{
			private object _dataObject;
			private readonly Point _mouseLocation;

			public DataObjectEventArgs(Point pt)
			{
				_mouseLocation = pt;
			}

			public Point MouseLocation
			{
				get { return _mouseLocation; }
			}

			public object DataObject
			{
				get { return _dataObject; }
				set { _dataObject = value; }
			}
		}

		public class DragDropEffectsEventArgs : EventArgs
		{
			private DragDropEffects _effects = DragDropEffects.None;

			public DragDropEffects DragDropEffects
			{
				get { return _effects; }
				set { _effects = value; }
			}
		}

		public class DragDropEventArgs : EventArgs
		{
			private readonly DragDropEffects _allowedEffects;
			private readonly DragDropEffects _effects;
			private readonly Point _mouseLocation;
			private readonly object _data;

			public DragDropEventArgs( object data, Point mouseLocation, DragDropEffects allowedEffects, DragDropEffects effects )
			{
				_data = data;
				_mouseLocation = mouseLocation;
				_allowedEffects = allowedEffects;
				_effects = effects;
			}

			public object Data
			{
				get { return _data; }
			}

			public Point MouseLocation
			{
				get { return _mouseLocation; }
			}

			public DragDropEffects Effects
			{
				get { return _effects; }
			}
			public DragDropEffects AllowedEffects
			{
				get { return _allowedEffects; }
			}
		}

		public class AllowDropEventArgs : EventArgs
		{
			public bool AllowDrop;
		}

		private static readonly Point NotDragging = new Point(-1, -1);

		private Point _startDragLocation;

		[Browsable(true)]
		[Category("Drag Drop")]
		[Description("Raised when the control needs to know what data should be dragged from the control.")]
		public event EventHandler<DataObjectEventArgs> GetDataObject;
		
		protected virtual void OnGetDataObject(DataObjectEventArgs e)
		{
			EventHandler<DataObjectEventArgs> handler = GetDataObject;
			
			if (handler != null)
				handler(this, e);
		}

		[Browsable(true)]
		[Category("Drag Drop")]
		[Description("Raised when the control needs to know what effects are allowed.")]
		public event EventHandler<DragDropEffectsEventArgs> GetDragEffects;

		protected virtual void OnGetDragEffects(DragDropEffectsEventArgs e)
		{
			EventHandler<DragDropEffectsEventArgs> handler = GetDragEffects;
			if (handler != null)
				handler(this, e);
		}
		
		[Browsable(true)]
		[Category("Drag Drop")]
		[Description("Raised when the drag and drop operation has completed.")]
		public event EventHandler<DragDropEventArgs> DragCompleted;

		protected virtual void OnDragCompleted(DragDropEventArgs e)
		{
			EventHandler<DragDropEventArgs> handler = DragCompleted;
			if (handler != null)
				handler(this, e);
		}

		[Browsable( true )]
		[Category( "Drag Drop" )]
		[Description( "Raised when a drag operation is over the control to determine whether the drop is allowed" )]
		public event EventHandler<AllowDropEventArgs> AllowDropOperation;

		protected virtual void OnAllowDropOperation(AllowDropEventArgs e)
		{
			EventHandler<AllowDropEventArgs> handler = AllowDropOperation;
			if (handler != null)
				handler(this, e);
		}

		protected override void OnMouseDown(MouseEventArgs e)
		{
			if (IsLeftMouseButton(e.Button))
			{
				_startDragLocation = e.Location;
			}
			base.OnMouseDown(e);
		}

		protected override void OnMouseUp(MouseEventArgs e)
		{
			if (IsLeftMouseButton(e.Button))
			{
				_startDragLocation = NotDragging;
			}
			base.OnMouseUp(e);
		}

		protected override void OnMouseMove(MouseEventArgs e)
		{
			base.OnMouseMove(e);
			if (IsLeftMouseButton(e.Button) && _startDragLocation != NotDragging)
			{
				//Last check to see if we've moved far enough to begin the drag/drop
				Size sz = new Size(Math.Abs(_startDragLocation.X - e.Location.X), Math.Abs(_startDragLocation.Y - e.Location.Y));
				if (sz.Height > SystemInformation.DragSize.Height || sz.Width > SystemInformation.DragSize.Width)
				{
					BeginDragDrop(_startDragLocation);
				}
			}
		}

		private void BeginDragDrop(Point startingPoint)
		{
			DataObjectEventArgs e = new DataObjectEventArgs(startingPoint);
			OnGetDataObject(e);
			object data = e.DataObject;
			if (data == null)
				return;
			DragDropEffectsEventArgs de = new DragDropEffectsEventArgs();
			OnGetDragEffects(de);
			DragDropEffects allowedEffects = de.DragDropEffects;
			DragDropEffects effects = DoDragDrop( data, allowedEffects );
			DragDropEffectsEventArgs finishedArgs = new DragDropEffectsEventArgs();
			finishedArgs.DragDropEffects = effects;
			DragDropEventArgs ddea = new DragDropEventArgs( data, startingPoint, allowedEffects, effects );
			OnDragCompleted( ddea );
		}

		private static bool IsLeftMouseButton( MouseButtons buttons )
		{
			return 0 != (buttons & MouseButtons.Left);
		}

		private void InitializeComponent()
		{
			SuspendLayout();
			ResumeLayout(false);
		}
	}
}

File: API/PageAndRDL/ReportWizard/VB.NET/ReportWizard/UI/DragDropListBox.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Namespace UI
	Public Class DragDropListBox
		Inherits ListBox
		Public Class DataObjectEventArgs
			Inherits EventArgs
			Private _dataObject As Object
			Private ReadOnly _mouseLocation As Point
			Public Sub New(ByVal pt As Point)
				_mouseLocation = pt
			End Sub
			Public ReadOnly Property MouseLocation() As Point
				Get
					Return _mouseLocation
				End Get
			End Property
			Public Property DataObject() As Object
				Get
					Return _dataObject
				End Get
				Set(ByVal value As Object)
					_dataObject = value
				End Set
			End Property
		End Class
		Public Class DragDropEffectsEventArgs
			Inherits EventArgs
			Private effects As DragDropEffects = System.Windows.Forms.DragDropEffects.None
			Public Property DragDropEffects() As DragDropEffects
				Get
					Return effects
				End Get
				Set(ByVal value As DragDropEffects)
					effects = Value
				End Set
			End Property
		End Class
		Public Class DragDropEventArgs
			Inherits EventArgs
			Private ReadOnly _allowedEffects As DragDropEffects
			Private ReadOnly _effects As DragDropEffects
			Private ReadOnly _mouseLocation As Point
			Private ReadOnly _data As Object
			Public Sub New(ByVal data As Object, ByVal mouseLocation As Point, ByVal allowedEffects As DragDropEffects, ByVal effects As DragDropEffects)
				Me._data = data
				Me._mouseLocation = mouseLocation
				Me._allowedEffects = allowedEffects
				Me._effects = effects
			End Sub
			Public ReadOnly Property Data() As Object
				Get
					Return _data
				End Get
			End Property
			Public ReadOnly Property MouseLocation() As Point
				Get
					Return _mouseLocation
				End Get
			End Property
			Public ReadOnly Property Effects() As DragDropEffects
				Get
					Return _effects
				End Get
			End Property
			Public ReadOnly Property AllowedEffects() As DragDropEffects
				Get
					Return _allowedEffects
				End Get
			End Property
		End Class
		Public Class AllowDropEventArgs
			Inherits EventArgs
			Public AllowDrop As Boolean
		End Class
		Private Shared ReadOnly NotDragging As New Point(-1, -1)
		Private startDragLocation As Point

		<Browsable(True)> _
		<Category("Drag Drop")>
		<Description("Raised when the control needs to know what data should be dragged from the control.")>
		Public Event GetDataObject As EventHandler(Of DataObjectEventArgs)

		Protected Overridable Sub OnGetDataObject(ByVal e As DataObjectEventArgs)
			'Dim handler As EventHandler(Of DataObjectEventArgs) = AddressOf GetDataObject
		   RaiseEvent GetDataObject(Me, e)
		End Sub

		<Browsable(True)> _
		<Category("Drag Drop")> _
		<Description("Raised when the control needs to know what effects are allowed.")> _
		Public Event GetDragEffects As EventHandler(Of DragDropEffectsEventArgs)

		Protected Overridable Sub OnGetDragEffects(ByVal e As DragDropEffectsEventArgs)
			RaiseEvent GetDragEffects(Me, e)
		End Sub

		<Browsable(True)> _
		<Category("Drag Drop")>
		<Description("Raised when the drag and drop operation has completed.")>
		Public Event DragCompleted As EventHandler(Of DragDropEventArgs)

		Protected Overridable Sub OnDragCompleted(ByVal e As DragDropEventArgs)
			RaiseEvent DragCompleted(Me, e)
		End Sub

		<Browsable(True)> _
		<Category("Drag Drop")> _
		<Description("Raised when a drag operation is over the control to determine whether the drop is allowed")> _
		Public Event AllowDropOperation As EventHandler(Of AllowDropEventArgs)


		Protected Overridable Sub OnAllowDropOperation(ByVal e As AllowDropEventArgs)
			RaiseEvent AllowDropOperation(Me, e)
		End Sub
		Protected Overloads Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			If IsLeftMouseButton(e.Button) Then
				startDragLocation = e.Location
			End If
			MyBase.OnMouseDown(e)
		End Sub
		Protected Overloads Overrides Sub OnMouseUp(ByVal e As MouseEventArgs)
			If IsLeftMouseButton(e.Button) Then
				startDragLocation = NotDragging
			End If
			MyBase.OnMouseUp(e)
		End Sub
		Protected Overloads Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)
			If IsLeftMouseButton(e.Button) AndAlso Not (startDragLocation).Equals(NotDragging) Then
				'' Last check to see if we've moved far enough to begin the drag/drop
				Dim sz As New Size(Math.Abs(startDragLocation.X - e.Location.X), Math.Abs(startDragLocation.Y - e.Location.Y))
				If sz.Height > SystemInformation.DragSize.Height OrElse sz.Width > SystemInformation.DragSize.Width Then
					BeginDragDrop(startDragLocation)
				End If
			End If
		End Sub
		Private Sub BeginDragDrop(ByVal startingPoint As Point)
			Dim e As New DataObjectEventArgs(startingPoint)
			OnGetDataObject(e)
			Dim data As Object = e.DataObject
			If (data) Is Nothing Then
				Return
			End If
			Dim de As New DragDropEffectsEventArgs()
			OnGetDragEffects(de)
			Dim allowedEffects As DragDropEffects = de.DragDropEffects
			Dim effects As DragDropEffects = DoDragDrop(data, allowedEffects)
			Dim finishedArgs As New DragDropEffectsEventArgs()
			finishedArgs.DragDropEffects = effects
			Dim ddea As New DragDropEventArgs(data, startingPoint, allowedEffects, effects)
			OnDragCompleted(ddea)
		End Sub
		Private Shared Function IsLeftMouseButton(ByVal buttons As MouseButtons) As Boolean
			Return 0 <> (buttons And MouseButtons.Left)
		End Function

		Private Sub InitializeComponent()
			Me.SuspendLayout()
			Me.ResumeLayout(False)

		End Sub
	End Class
End Namespace
